File: 295.Find_Median_from_Data_Stream.cc
#include <iostream>
#include <cassert>
#include <queue>
class MedianFinder {
public:
    MedianFinder() {}
    
    void addNum(int num) {
        if (maxHeap.empty() || num <= maxHeap.top()) {
            maxHeap.push(num);
        } else {
            minHeap.push(num);
        }

        if (maxHeap.size() > minHeap.size() + 1) {
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        } else if (maxHeap.size() < minHeap.size()) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }
    
    double findMedian() {
        if (maxHeap.size() == minHeap.size()) {
            return (double)(maxHeap.top() + minHeap.top()) / 2;
        } else {
            return (double)maxHeap.top();
        }
    }
private:
    std::priority_queue<int> maxHeap;
    std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;
    int numVals = 0;
};

/*
    Split the values into two heaps, one max and one min
    If odd num of values, the val at the top of the max heap will be the median
    If even, the median is the average of the top of the max and min heap

    eg. 1 2 3
    max 1 2
    min 3
    median = 2

    eg 1 2 3 4
    max 1 2
    min 4 3
    median = 2 + 3 / 2 = 2.5 
*/

int main() {
    MedianFinder mf;

    mf.addNum(1);
    mf.addNum(2);
    assert(mf.findMedian() == 1.5);
    mf.addNum(3);
    assert(mf.findMedian() == 2.0);

    std::cout << "\033[32mAll tests passed!\033[0m\n";
    return 0;
}

File: Python solutions/295. Find Median from Data Stream.py
import heapq

import heapq

class MedianFinder:

    def __init__(self):
        self.maxHeap = []
        self.minHeap = []

    def addNum(self, num: int) -> None:
        if not self.maxHeap or num <= -self.maxHeap[0]:
            heapq.heappush(self.maxHeap, -num)
        else:
            heapq.heappush(self.minHeap, num)

        # Rebalance heaps if necessary to maintain the property that maxHeap is either equal in size to minHeap or has one extra element
        if len(self.maxHeap) > len(self.minHeap) + 1:
            heapq.heappush(self.minHeap, -heapq.heappop(self.maxHeap))
        elif len(self.maxHeap) < len(self.minHeap):
            heapq.heappush(self.maxHeap, -heapq.heappop(self.minHeap))

    def findMedian(self) -> float:
        if len(self.maxHeap) == len(self.minHeap):
            return (-self.maxHeap[0] + self.minHeap[0]) / 2.0
        else:
            return -self.maxHeap[0]



"""
To find median, use a max heap to store the lower half of nums
    use min heap to store the higher half of nums
Ex
    Max heap: 1  2 3 4 5
    Min heap: 10 9 8 7 6
    --------------------
    Median: (5 + 6) / 2 = 11 / 2 = 5.5
    That way the median can be found in O(1) time by taking the top of each


"""
        

assert Solution(). == , "Test failed!"

print("\033[32mAll tests passed!\033[0m\n")